eclair-core/src/main/scala/fr/acinq/eclair/wire/Onion.scala

```diff
@@ -18,48 +18,136 @@ package fr.acinq.eclair.wire
 
 import fr.acinq.bitcoin.ByteVector32
 import fr.acinq.eclair.crypto.Sphinx
-import fr.acinq.eclair.{CltvExpiry, MilliSatoshi, ShortChannelId}
-import scodec.bits.{BitVector, ByteVector}
+import fr.acinq.eclair.wire.CommonCodecs._
+import fr.acinq.eclair.wire.OnionTlv._
+import fr.acinq.eclair.wire.TlvCodecs._
+import fr.acinq.eclair.{CltvExpiry, MilliSatoshi, ShortChannelId, UInt64}
+import scodec.bits.{BitVector, ByteVector, HexStringSyntax}
 import scodec.codecs._
 import scodec.{Codec, DecodeResult, Decoder}
 
 /**
  * Created by t-bast on 05/07/2019.
  */
 
+/**
+ * Tlv types used inside onion messages.
+ */
+sealed trait OnionTlv extends Tlv
+
 case class OnionRoutingPacket(version: Int,
                               publicKey: ByteVector,
                               payload: ByteVector,
                               hmac: ByteVector32)
 
-case class PerHopPayload(shortChannelId: ShortChannelId,
-                         amtToForward: MilliSatoshi,
-                         outgoingCltvValue: CltvExpiry)
+case class OnionForwardInfo(shortChannelId: ShortChannelId,
+                            amtToForward: MilliSatoshi,
+                            outgoingCltvValue: CltvExpiry)
+
+case class OnionPaymentInfo(amount: MilliSatoshi, cltvExpiry: CltvExpiry)
+
+case class OnionPerHopPayload(payload: Either[TlvStream[OnionTlv], OnionForwardInfo]) {
```

Comment:
Have you considered using a proper trait hierarchy?

```scala
sealed trait Payload

sealed trait LegacyPayload extends Payload
sealed trait TlvPayload extends Payload

sealed trait RelayPayload extends Payload
sealed trait FinalPayload extends Payload

case class RelayLegacy(...) extends LegacyPayload with RelayPayload
case class FinalLegacy(...) extends LegacyPayload with FinalPayload
case class RelayTlv(...) extends TlvPayload with RelayPayload
case class FinalTlv(...) extends TlvPayload with FinalPayload
```

Reply:
This is more verbose, but you're right that it will likely make the code more explicit and allow us to separate legacy and tlv more cleanly. I'll investigate that.

Reply:
This is much better, thanks for the suggestion.
We are really going to have different code-paths for tlv vs legacy and relay vs final, so it really makes sense to separate them explicitly at the type level.